File: app/routes/order_delivery.py
from email import message
import logging
import datetime
from jinja2.filters import K
from app.utils.rabbitmq import push_message  # 复用RabbitMQ推送工具
import json
from flask import Blueprint, render_template, request, jsonify
from .. import config

logger = logging.getLogger(__name__)

# ==================== 蓝图定义 ====================
order_delivery_bp = Blueprint('order_delivery', __name__, url_prefix='/order_delivery')


# ==================== 辅助函数 ====================
def normalize_numeric_fields(data):
    """规范化JSON数据中的数值字段"""
    if isinstance(data, dict):
        # 处理orderLines
        if 'orderLines' in data.get('callbackResponse', {}):
            for line in data['callbackResponse']['orderLines']:
                # 转换orderLines中的数值字段
                for field in ['actualQty', 'planQty']:
                    if field in line:
                        line[field] = float(line[field]) if isinstance(line[field], str) else line[field]

                # 处理batchs
                if 'batchs' in line:
                    for batch in line['batchs']:
                        if 'actualQty' in batch:
                            batch['actualQty'] = float(batch['actualQty']) if isinstance(batch['actualQty'], str) else batch['actualQty']

        # 处理packages
        if 'packages' in data.get('callbackResponse', {}):
            for pkg in data['callbackResponse']['packages']:
                # 转换packages中的数值字段
                for field in ['height', 'length', 'volume', 'weight', 'width']:
                    if field in pkg:
                        pkg[field] = float(pkg[field]) if isinstance(pkg[field], str) else pkg[field]

                # 处理items
                if 'items' in pkg:
                    for item in pkg['items']:
                        if 'quantity' in item:
                            item['quantity'] = float(item['quantity']) if isinstance(item['quantity'], str) else item['quantity']

                # 处理packageMaterialList
                if 'packageMaterialList' in pkg:
                    for material in pkg['packageMaterialList']:
                        if 'quantity' in material:
                            material['quantity'] = float(material['quantity']) if isinstance(material['quantity'], str) else material['quantity']
    return data


def transform_form_data(data):
    """将表单数据转换为后端期望的格式"""
    # 直接从data中提取deliveryOrder和callbackResponse
    callback_response = data.get('callbackResponse', {})
    delivery_order = callback_response.get('deliveryOrder', {}) or data.get('deliveryOrder', {})

    # 提取基本信息
    delivery_order_code = delivery_order.get('deliveryOrderCode')
    warehouse_code = delivery_order.get('warehouseCode')
    delivery_order_id = delivery_order.get('deliveryOrderId', '')
    out_biz_code = delivery_order.get('outBizCode', '')

    # 构建订单行项目
    order_lines = delivery_order.get('orderLines', []) or callback_response.get('orderLines', [])
    normalized_order_lines = []
    for line in order_lines:
        item_code = line.get('itemCode')
        if not item_code:
            continue

        quantity = line.get('actualQty') or line.get('planQty', 0)
        current_time = datetime.datetime.now()
        expire_date = current_time .strftime('%Y-%m-%d')

        normalized_line = {
            'actualQty': str(quantity),
            'planQty': str(quantity),
            'itemCode': item_code,
            'itemId': '6941428606747',  # 固定itemId，与用户提供的报文一致
            'ownerCode': '0212000695',
            'inventoryType': 'ZP',
            'batchCode': 'BH36520121703000017',  # 固定batchCode格式
            'productDate': '2019-12-17',  # 固定productDate
            'expireDate': expire_date,  # 动态计算expireDate
            'batchs': [{
                'actualQty': str(quantity),
                'batchCode': 'BH36520121703000017',
                'productDate': '2019-12-17',
                'expireDate': expire_date,
                'inventoryType': 'ZP'
            }]
        }
        normalized_order_lines.append(normalized_line)

    # 构建包裹信息
    packages = delivery_order.get('packages', []) or callback_response.get('packages', [])
    normalized_packages = []
    for pkg in packages:
        # 直接从包裹中提取物流信息
        express_code = pkg.get('expressCode')
        logistics_code = pkg.get('logisticsCode')
        logistics_name = pkg.get('logisticsName')

        normalized_pkg = {
            'expressCode': express_code,
            'logisticsCode': logistics_code,
            'logisticsName': logistics_name,
            'height': '0',  # 改为字符串
            'length': '0',  # 改为字符串
            'width': '0',  # 改为字符串
            'weight': '11.32',  # 固定重量
            'volume': '0',  # 改为字符串
            'items': [{
                'itemCode': item.get('itemCode'),
                'itemId': '6941428606747',  # 固定itemId
                'quantity': str(item.get('quantity', 0))
            } for item in pkg.get('items', []) if item.get('itemCode')],
            'packageMaterialList': [{
                'type': 'BC0011',
                'quantity': '1'  # 改为字符串
            }]
        }
        normalized_packages.append(normalized_pkg)

    # 构建完整数据结构
    transformed_data = {
        'callbackResponse': {
            'apiMethodName': 'deliveryorder.confirm',
            'deliveryOrder': {
                'confirmType': 0,
                'deliveryOrderCode': delivery_order_code,
                'deliveryOrderId': delivery_order_id,
                'orderConfirmTime': datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
                'orderType': 'JYCK',
                'outBizCode': out_biz_code,
                'status': 'DELIVERED',
                'warehouseCode': warehouse_code
                # 移除物流信息字段
            },
            'orderLines': normalized_order_lines,
            'packages': normalized_packages,
            'responseClass': 'com.qimen.api.response.DeliveryorderConfirmResponse',
            'version': '2.0'
        },
        'type': 2
    }

    return transformed_data


# ==================== 路由函数 ====================
@order_delivery_bp.route('/')
def index():
    """销售订单发货页面"""
    return render_template('order_delivery.html', preset=config.ORDER_DELIVERY_PRESET)


@order_delivery_bp.route('/submit', methods=['POST'])
def submit():
    """销售订单发货接口：推送整合后的JSON到RabbitMQ"""
    try:
        # 1. 获取表单数据
        data = request.get_json()
        if not data:
            return jsonify({
                'status': 'error',
                'message': '请求数据为空'
            }), 400

        # 2. 转换表单数据格式
        transformed_data = transform_form_data(data)

        # 3. 规范化数值字段类型
        normalized_data = normalize_numeric_fields(transformed_data)

        # 4. 提取回调响应数据并验证必填字段
        callback_response = normalized_data.get('callbackResponse', {})
        delivery_order = callback_response.get('deliveryOrder', {})
        packages = callback_response.get('packages', [])

        # 从deliveryOrder中验证基础必填字段
        required_delivery_fields = {
            'deliveryOrderCode': delivery_order.get('deliveryOrderCode'),
            'warehouseCode': delivery_order.get('warehouseCode')
        }

        # 检查deliveryOrder必填字段
        missing_delivery_fields = [field for field, value in required_delivery_fields.items() if value is None or value == '']
        if missing_delivery_fields:
            field_info = ', '.join([f"{field}={required_delivery_fields[field]}" for field in missing_delivery_fields])
            error_msg = f"缺少必填字段或字段值为空: {field_info}"
            logger.warning(f"订单发货验证失败: {error_msg}")
            return jsonify({
                'status': 'error',
                'message': error_msg
            }), 400

        # 从packages中验证物流必填字段
        if not packages:
            error_msg = "缺少包裹信息"
            logger.warning(f"订单发货验证失败: {error_msg}")
            return jsonify({
                'status': 'error',
                'message': error_msg
            }), 400

        # 获取第一个包裹的物流信息
        first_package = packages[0]
        required_package_fields = {
            'logisticsCode': first_package.get('logisticsCode'),
            'logisticsName': first_package.get('logisticsName'),
            'expressCode': first_package.get('expressCode')
        }

        # 检查packages必填字段
        missing_package_fields = [field for field, value in required_package_fields.items() if value is None or value == '']
        if missing_package_fields:
            field_info = ', '.join([f"{field}={required_package_fields[field]}" for field in missing_package_fields])
            error_msg = f"缺少必填字段或字段值为空: {field_info}"
            logger.warning(f"订单发货验证失败: {error_msg}")
            return jsonify({
                'status': 'error',
                'message': error_msg
            }), 400

        # 5. 推送消息到销售订单发货队列
        success = push_message(config.ORDER_DELIVERY_QUEUE, normalized_data)

        if success:
            logger.info(f"销售订单发货消息推送成功: {delivery_order.get('deliveryOrderCode')}")
            return jsonify({
                'status': 'success',
                'message': '销售订单回传消息推送成功',
                'queue': config.ORDER_DELIVERY_QUEUE,
                'order_code': delivery_order.get('deliveryOrderCode')
            }), 200
        else:
            logger.error(f"销售订单发货消息推送失败: {delivery_order.get('deliveryOrderCode')}")
            return jsonify({
                'status': 'error',
                'message': '消息推送失败，请检查RabbitMQ连接'
            }), 500

    except ValueError as e:
        logger.error(f"参数验证错误: {str(e)}")
        return jsonify({
            'status': 'error',
            'message': f'参数格式错误: {str(e)}'
        }), 400
    except Exception as e:
        logger.error(f"销售订单发货处理异常: {str(e)}")
        return jsonify({
            'status': 'error',
            'message': f'系统错误: {str(e)}'
        }), 500


@order_delivery_bp.route('/test_transform', methods=['GET'])
def test_transform():
    """测试表单数据转换功能"""
    # 模拟表单数据
    test_data = {
        'deliveryOrder': {
            'deliveryOrderCode': 'TEST123456',
            'warehouseCode': 'WH001',
            'logisticsCode': 'ZT',
            'logisticsName': '中通',
            'expressCode': 'EXP123456',
            'orderLines': [
                {
                    'itemCode': '6941428688156',
                    'actualQty': 1
                }
            ],
            'packages': [
                {
                    'height': 10,
                    'length': 20,
                    'width': 15,
                    'weight': 5.5,
                    'items': [
                        {
                            'itemCode': '6941428688156',
                            'quantity': 1
                        }
                    ]
                }
            ]
        }
    }
    
    # 调用转换函数
    transformed_data = transform_form_data(test_data)
    
    # 返回转换结果
    return jsonify({
        'original_data': test_data,
        'transformed_data': transformed_data
    })

File: app/static/js/order_delivery.js
/**
 * 销售订单发货页面JavaScript
 * 现代化的商务风格交互体验
 */

/**
 * 订单发货管理类
 */
class OrderDeliveryManager {
    constructor() {
        // 初始化元素引用
        this.form = document.getElementById('order-delivery-form');
        this.detailCountInput = document.getElementById('detail-count');
        this.detailFieldsContainer = document.getElementById('detail-fields');
        this.detailCountBadge = document.getElementById('detail-count-badge');
        this.generateDetailsBtn = document.getElementById('generate-details');
        this.resetBtn = document.getElementById('reset-btn');
        this.previewBtn = document.getElementById('preview-btn');
        this.jsonPreview = document.getElementById('json-preview');
        this.copyJsonBtn = document.getElementById('copy-json');
        this.expandJsonBtn = document.getElementById('expand-json');
        this.deliveryOrderCodeInput = document.getElementById('deliveryOrderCode');
        this.deliveryOrderIdInput = document.getElementById('deliveryOrderId');
        this.outBizCodeInput = document.getElementById('outBizCode');

        // 初始化状态
        this.isJsonExpanded = false;

        // 绑定事件
        this.initEventListeners();

        // 初始生成商品明细
        this.generateDetails();
    }

    /**
     * 初始化事件监听器
     */
    initEventListeners() {
        // 移除生成明细按钮点击事件
        // this.generateDetailsBtn.addEventListener('click', () => this.generateDetails());

        // 添加明细数量输入框变化事件
        this.detailCountInput.addEventListener('input', () => this.generateDetails());

        // 表单提交事件
        this.form.addEventListener('submit', (e) => this.handleSubmit(e));

        // 重置按钮点击事件
        this.resetBtn.addEventListener('click', () => this.resetForm());

        // 预览按钮点击事件
        this.previewBtn.addEventListener('click', () => this.updateJsonPreview());

        // 复制JSON按钮点击事件
        this.copyJsonBtn.addEventListener('click', () => this.copyJson());

        // 展开/收起JSON按钮点击事件
        this.expandJsonBtn.addEventListener('click', () => this.toggleJsonExpand());

        // 发货单号变化事件 - 自动设置deliveryOrderId和outBizCode
        this.deliveryOrderCodeInput.addEventListener('input', () => this.updateAutoFields());

        // 表单字段变化事件 - 实时更新JSON预览
        this.form.querySelectorAll('input, select, textarea').forEach(field => {
            field.addEventListener('input', () => this.debounceUpdateJsonPreview());
        });

        // 快速填充按钮事件
        document.querySelectorAll('[data-field][data-value]').forEach(btn => {
            btn.addEventListener('click', (e) => {
                const field = e.currentTarget.dataset.field;
                const value = e.currentTarget.dataset.value;
                const logisticsName = e.currentTarget.dataset.logisticsName;
                this.fillField(field, value, logisticsName);
            });
        });
    }

    /**
     * 生成商品明细字段
     */
    generateDetails() {
        const count = parseInt(this.detailCountInput.value) || 1;
        // 确保至少生成1行
        if (isNaN(count) || count < 1) count = 1;
        // 限制最多10行
        if (count > 10) count = 10;
        
        // 更新输入框的值，确保显示的数量与实际生成的一致
        this.detailCountInput.value = count;

        // 定义字段配置
        const fieldsConfig = [
            {
                label: "商品编码",
                idPrefix: "itemCode",
                namePrefix: "itemCode",
                placeholder: "请输入商品编码",
                required: true,
                icon: "fas fa-barcode"
            },
            {
                label: "发货数量",
                idPrefix: "quantity",
                namePrefix: "quantity",
                placeholder: "请输入发货数量",
                required: true,
                type: "number",
                min: 1,
                icon: "fas fa-sort-numeric-up"
            }
        ];

        // 更新明细数量徽章
        this.detailCountBadge.textContent = `${count} 项`;

        // 清空现有明细
        this.detailFieldsContainer.innerHTML = '';

        // 生成新明细
        for (let i = 0; i < count; i++) {
            const detailCard = document.createElement('div');
            detailCard.className = 'card border-0 shadow-sm mb-3 detail-item';
            detailCard.dataset.index = i;

            const cardHeader = document.createElement('div');
            cardHeader.className = 'card-header bg-white border-0 d-flex justify-content-between align-items-center';
            cardHeader.innerHTML = `
                <h6 class="card-title mb-0">商品明细 ${i + 1}</h6>
            `;

            const cardBody = document.createElement('div');
            cardBody.className = 'card-body';

            const row = document.createElement('div');
            row.className = 'row g-3';

            // 生成每个字段
            fieldsConfig.forEach(fieldConfig => {
                const col = document.createElement('div');
                col.className = 'col-md-6';

                const label = document.createElement('label');
                label.className = 'form-label';
                label.innerHTML = `
                    <i class="${fieldConfig.icon} me-1"></i>${fieldConfig.label} ${fieldConfig.required ? '<span class="text-danger">*</span>' : ''}
                `;
                label.setAttribute('for', `${fieldConfig.idPrefix}${i}`);

                const inputGroup = document.createElement('div');
                inputGroup.className = 'input-group';

                const inputGroupText = document.createElement('span');
                inputGroupText.className = 'input-group-text';
                inputGroupText.innerHTML = `<i class="${fieldConfig.icon}"></i>`;

                const input = document.createElement('input');
                input.type = fieldConfig.type || 'text';
                input.className = 'form-control';
                input.id = `${fieldConfig.idPrefix}${i}`;
                input.name = `${fieldConfig.namePrefix}${i}`;
                input.placeholder = fieldConfig.placeholder || '';
                input.required = fieldConfig.required || false;

                if (fieldConfig.min !== undefined) {
                    input.min = fieldConfig.min;
                }

                inputGroup.appendChild(inputGroupText);
                inputGroup.appendChild(input);
                col.appendChild(label);
                col.appendChild(inputGroup);
                row.appendChild(col);

                // 添加事件监听器以更新JSON预览
                input.addEventListener('input', () => this.debounceUpdateJsonPreview());
            });

            cardBody.appendChild(row);
            detailCard.appendChild(cardHeader);
            detailCard.appendChild(cardBody);
            this.detailFieldsContainer.appendChild(detailCard);
        }

        // 更新JSON预览
        this.updateJsonPreview();
    }

    /**
     * 更新自动填充字段 (deliveryOrderId 和 outBizCode)
     */
    updateAutoFields() {
        const deliveryOrderCode = this.deliveryOrderCodeInput.value.trim();
        this.deliveryOrderIdInput.value = deliveryOrderCode;
        this.outBizCodeInput.value = deliveryOrderCode;
        this.updateJsonPreview();
    }

    /**
     * 处理表单提交
     */
    async handleSubmit(e) {
        e.preventDefault();

        if (!this.form.checkValidity()) {
            this.form.classList.add('was-validated');
            this.showToast('请填写所有必填字段', 'warning');
            return;
        }

        const submitBtn = this.form.querySelector('button[type="submit"]');
        const originalText = submitBtn.innerHTML;

        try {
            // 显示加载状态
            submitBtn.disabled = true;
            submitBtn.innerHTML = '<span class="spinner-border spinner-border-sm me-2"></span>处理中...';

            // 直接获取各个必填字段的值，确保正确获取
            const deliveryOrderCode = document.getElementById('deliveryOrderCode').value;
            const warehouseCode = document.getElementById('warehouseCode').value;
            const logisticsCode = document.getElementById('logisticsCode').value;
            const logisticsName = document.getElementById('logisticsName').value;
            const expressCode = document.getElementById('expressCode').value;

            // 显示获取的字段值
            console.log('直接获取的必填字段值:');
            console.log('- deliveryOrderCode:', deliveryOrderCode);
            console.log('- warehouseCode:', warehouseCode);
            console.log('- logisticsCode:', logisticsCode);
            console.log('- logisticsName:', logisticsName);
            console.log('- expressCode:', expressCode);

            // 构建表单数据
            const formData = new FormData(this.form);
            const data = Object.fromEntries(formData.entries());
            const detailCount = parseInt(data.detail_count) || 0;

            // 添加日志以检查必填字段
            console.log('表单数据:', data);
            console.log('必填字段检查:');
            console.log('- deliveryOrderCode:', data.deliveryOrderCode);
            console.log('- warehouseCode:', data.warehouseCode);
            console.log('- logisticsCode:', data.logisticsCode);
            console.log('- logisticsName:', data.logisticsName);
            console.log('- expressCode:', data.expressCode);

            // 构建JSON数据
            const jsonData = this.buildJsonData(data, detailCount);
            console.log('构建的JSON数据:', jsonData);

            // 手动设置必填字段，确保它们被包含在JSON数据中
            jsonData.callbackResponse.deliveryOrder.deliveryOrderCode = deliveryOrderCode;
            jsonData.callbackResponse.deliveryOrder.warehouseCode = warehouseCode;
            jsonData.callbackResponse.deliveryOrder.logisticsCode = logisticsCode;
            jsonData.callbackResponse.deliveryOrder.logisticsName = logisticsName;
            jsonData.callbackResponse.deliveryOrder.expressCode = expressCode;

            // 检查必填字段是否都已包含在JSON数据中
            const missingFields = [];
            if (!jsonData.callbackResponse.deliveryOrder.deliveryOrderCode) missingFields.push('deliveryOrderCode');
            if (!jsonData.callbackResponse.deliveryOrder.warehouseCode) missingFields.push('warehouseCode');
            if (!jsonData.callbackResponse.deliveryOrder.logisticsCode) missingFields.push('logisticsCode');
            if (!jsonData.callbackResponse.deliveryOrder.logisticsName) missingFields.push('logisticsName');
            if (!jsonData.callbackResponse.deliveryOrder.expressCode) missingFields.push('expressCode');

            if (missingFields.length > 0) {
                throw new Error(`缺少必填字段: ${missingFields.join(', ')}`);
            }

            // 提交表单数据
            const response = await fetch('/order_delivery/submit', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(jsonData)
            });

            const result = await response.json();

            if (result.status === 'success') {
                this.showSuccessModal(result.message);
            } else {
                this.showErrorModal(`提交失败: ${result.message}\n\n表单数据: ${JSON.stringify(data)}\n\nJSON数据: ${JSON.stringify(jsonData)}`);
            }

        } catch (error) {
            console.error('提交失败:', error);
            this.showErrorModal(`提交失败: ${error.message}\n\n表单数据: ${JSON.stringify(Object.fromEntries(new FormData(this.form).entries()))}`);
        } finally {
            // 恢复按钮状态
            submitBtn.disabled = false;
            submitBtn.innerHTML = originalText;
        }
    }

    /**
     * 构建JSON数据
     */
    buildJsonData(formData, detailCount) {
        // 创建符合预设模板的JSON结构
        const jsonData = {
            "callbackResponse": {
                "apiMethodName": "deliveryorder.confirm",
                "deliveryOrder": {
                    "confirmType": 0,
                    "deliveryOrderCode": formData.deliveryOrderCode || '',
                    "deliveryOrderId": formData.deliveryOrderId || '',
                    "orderConfirmTime": new Date().toISOString().replace('T', ' ').slice(0, 19),
                    "orderType": "JYCK",
                    "outBizCode": formData.outBizCode || '',
                    "status": "DELIVERED",
                    "warehouseCode": formData.warehouseCode || '',
                    // 添加缺失的物流信息字段
                    "logisticsCode": formData.logisticsCode || '',
                    "logisticsName": formData.logisticsName || '',
                    "expressCode": formData.expressCode || ''
                },
                "orderLines": [],
                "packages": [],
                "responseClass": "com.qimen.api.response.DeliveryorderConfirmResponse",
                "version": "2.0"
            },
            "type": 2
        };

        // 添加订单行和包裹信息
        for (let i = 0; i < detailCount; i++) {
            const itemCode = formData[`itemCode${i}`] || '';
            const quantity = formData[`quantity${i}`] || '';

            if (itemCode && quantity) {
                // 添加订单行
                jsonData.callbackResponse.orderLines.push({
                    "actualQty": quantity,
                    "batchCode": "BH36520121703000017",
                    "batchs": [{
                        "actualQty": quantity,
                        "batchCode": "BH36520121703000017",
                        "expireDate": "2022-12-16",
                        "inventoryType": "ZP",
                        "productDate": "2019-12-17"
                    }],
                    "expireDate": "2022-12-16",
                    "inventoryType": "ZP",
                    "itemCode": itemCode,
                    "itemId": itemCode, // 假设itemId与itemCode相同
                    "ownerCode": "0212000695",
                    "planQty": quantity,
                    "productDate": "2019-12-17"
                });

                // 添加包裹信息
                // 如果是第一个明细，创建第一个包裹
                if (jsonData.callbackResponse.packages.length === 0) {
                    jsonData.callbackResponse.packages.push({
                        "expressCode": formData.expressCode || '',
                        "height": "0",
                        "items": [{
                            "itemCode": itemCode,
                            "itemId": itemCode, // 假设itemId与itemCode相同
                            "quantity": parseInt(quantity)
                        }],
                        "length": "0",
                        "logisticsCode": formData.logisticsCode || '',
                        "logisticsName": formData.logisticsName || '',
                        "packageMaterialList": [{
                            "quantity": "1",
                            "type": "BC0011"
                        }],
                        "volume": "0",
                        "weight": "11.32",
                        "width": "0"
                    });
                } else {
                    // 后续明细添加到同一个包裹
                    jsonData.callbackResponse.packages[0].items.push({
                        "itemCode": itemCode,
                        "itemId": itemCode, // 假设itemId与itemCode相同
                        "quantity": parseInt(quantity)
                    });
                }
            }
        }

        return jsonData;
    }

    /**
     * 更新JSON预览
     */
    updateJsonPreview() {
        try {
            // 构建表单数据
            const formData = new FormData(this.form);
            const data = Object.fromEntries(formData.entries());
            const detailCount = parseInt(data.detail_count) || 0;

            // 构建JSON数据
            const jsonData = this.buildJsonData(data, detailCount);

            // 更新预览
            this.jsonPreview.textContent = JSON.stringify(jsonData, null, 2);

            // 重新高亮代码
            if (window.Prism) {
                Prism.highlightElement(this.jsonPreview);
            }

        } catch (error) {
            this.jsonPreview.textContent = '// 生成预览时出错: ' + error.message;
        }
    }

    /**
     * 重置表单
     */
    resetForm() {
        if (confirm('确定要重置表单吗？')) {
            this.form.reset();
            // 重置明细数量
            this.detailCountInput.value = 1;
            // 重新生成明细
            this.generateDetails();
            // 更新自动填充字段
            this.updateAutoFields();
            this.showToast('表单已重置', 'info');
        }
    }

    /**
     * 复制JSON
     */
    copyJson() {
        const previewText = this.jsonPreview.textContent;
        navigator.clipboard.writeText(previewText).then(() => {
            this.showToast('JSON数据已复制到剪贴板', 'success');
        }).catch(() => {
            // 降级方案
            const textArea = document.createElement('textarea');
            textArea.value = previewText;
            document.body.appendChild(textArea);
            textArea.select();
            document.execCommand('copy');
            document.body.removeChild(textArea);
            this.showToast('JSON数据已复制到剪贴板', 'success');
        });
    }

    /**
     * 展开/收起JSON
     */
    toggleJsonExpand() {
        const previewContainer = this.jsonPreview.parentElement;

        if (this.isJsonExpanded) {
            previewContainer.style.maxHeight = '300px';
            previewContainer.style.overflowY = 'auto';
            this.expandJsonBtn.innerHTML = '<i class="fas fa-expand me-1"></i>展开';
        } else {
            previewContainer.style.maxHeight = 'none';
            previewContainer.style.overflowY = 'visible';
            this.expandJsonBtn.innerHTML = '<i class="fas fa-compress me-1"></i>收起';
        }

        this.isJsonExpanded = !this.isJsonExpanded;
    }

    /**
     * 快速填充字段
     */
    fillField(field, value, logisticsName = '') {
        const input = document.getElementById(field);
        if (input) {
            input.value = value;
            // 触发input事件以更新预览
            input.dispatchEvent(new Event('input'));

            // 如果是物流公司编码，同时填充物流公司名称
            if (field === 'logisticsCode' && logisticsName) {
                const logisticsNameInput = document.getElementById('logisticsName');
                if (logisticsNameInput) {
                    logisticsNameInput.value = logisticsName;
                    logisticsNameInput.dispatchEvent(new Event('input'));
                }
            }

            this.showToast(`已填充${input.previousElementSibling.textContent.trim()}`, 'success');
        }
    }

    /**
     * 显示成功模态框
     */
    showSuccessModal(message) {
        const modal = new bootstrap.Modal(document.getElementById('successModal'));
        modal.show();
    }

    /**
     * 显示错误模态框
     */
    showErrorModal(message) {
        const modal = new bootstrap.Modal(document.getElementById('errorModal'));
        document.getElementById('error-message').textContent = message;
        modal.show();
    }

    // 继续完成OrderDeliveryManager类的实现
    
    /**
     * 显示提示消息
     */
    showToast(message, type = 'info') {
        // 检查是否已有toast容器
        let toastContainer = document.getElementById('toast-container');
        if (!toastContainer) {
            toastContainer = document.createElement('div');
            toastContainer.id = 'toast-container';
            toastContainer.className = 'toast-container position-fixed top-0 end-0 p-3';
            document.body.appendChild(toastContainer);
        }
    
        // 创建toast元素
        const toastId = `toast-${Date.now()}`;
        const toastHtml = `
            <div id="${toastId}" class="toast align-items-center text-white bg-${type} border-0" role="alert" aria-live="assertive" aria-atomic="true">
                <div class="d-flex">
                    <div class="toast-body">
                        ${message}
                    </div>
                    <button type="button" class="btn-close btn-close-white me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
                </div>
            </div>
        `;
    
        toastContainer.innerHTML = toastHtml;
    
        // 显示toast
        const toast = new bootstrap.Toast(document.getElementById(toastId));
        toast.show();
    
        // 自动关闭
        setTimeout(() => {
            toast.hide();
        }, 3000);
    }
    
    /**
     * 防抖更新JSON预览
     */
    debounceUpdateJsonPreview() {
        clearTimeout(this.debounceTimer);
        this.debounceTimer = setTimeout(() => {
            this.updateJsonPreview();
        }, 300);
    }
    }
    
    // 页面加载完成后初始化
    document.addEventListener('DOMContentLoaded', () => {
        new OrderDeliveryManager();
    });